external/storm-autocreds/src/main/java/org/apache/storm/common/AbstractAutoCreds.java

```diff
@@ -0,0 +1,250 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.storm.common;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.math3.util.Pair;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.security.Credentials;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+import org.apache.hadoop.security.token.TokenIdentifier;
+import org.apache.storm.security.INimbusCredentialPlugin;
+import org.apache.storm.security.auth.IAutoCredentials;
+import org.apache.storm.security.auth.ICredentialsRenewer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.security.auth.Subject;
+import javax.xml.bind.DatatypeConverter;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The base class that for auto credential plugins that abstracts out some of the common functionality.
+ */
+public abstract class AbstractAutoCreds implements IAutoCredentials, ICredentialsRenewer, INimbusCredentialPlugin {
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractAutoCreds.class);
+    public static final String CONFIG_KEY_RESOURCES = "resources";
+
+    private List<String> configKeys = new ArrayList<>();
+    private Map<String, Map<String, Object>> configMap = new HashMap<>();
+
+    @Override
+    public void prepare(Map conf) {
+        doPrepare(conf);
+        String configKeyString = getConfigKeyString();
+        if (conf.containsKey(configKeyString)) {
+            configKeys.addAll((List<String>) conf.get(configKeyString));
+            for (String key : configKeys) {
+                if (conf.containsKey(key)) {
+                    Map<String, Object> config = (Map<String, Object>) conf.get(key);
+                    configMap.put(key, config);
+                    LOG.info("configKey = {}, config = {}", key, config);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void populateCredentials(Map<String, String> credentials, Map conf) {
+        try {
+            if (configKeys != null) {
```

Comment:
configKeys is initialised to ArrayList. we need to use configKeys.isEmpty() check here.